### DISCUSSION TITLE
How will the spread operator from collection expressions mix with range objects?

### DISCUSSION THREAD

**Discussion by Captain-Quack** [dotnet/csharplang#7405] [LDM Notes]
I find it a bit worrying that both the range object and spread operator use `..`. 

For example, take this code:
```csharp
var range = 1..5;
var newArray = [..range]; 
```
Currently, this generates "`error CS0029: Cannot implicitly convert type 'System.Range' to 'System.Index'`". If the compier (or new programmer) sees this as a range operator, that could cause some confusion. 

Multidimensional collections could act strangely too:
```csharp
int[,] numbers = new[,] {
     {1, 2, 3, 4, 5},
     {1, 2, 3, 4, 5},
     {1, 2, 3, 4, 5},
};

Range range = 1..3;
var newArray = numbers[..range];
 ``` 
Here, `newArray` could:
- Access the 2nd and 3rd Array
- Access the 3rd element of the 2nd array
- Be invalid syntax, because `0..1..3` is not valid syntax.

Finally, this:
```csharp
[/* put any list here */][....]
```
Nobody in their right mind would type this, but...

I think the spread operator is a great idea, but it could bring up a lot of questions/issues in the future.

---
Related:
#7351: Also raises a question of how empty/null ranges will be handled **(`..` is valid syntax, and prints as `0..^0`!)**

**Comment by CyrusNajmabadi** [comment]
> How will the spread operator from collection expressions mix with range objects?

If 'range' becomes iterable then this will work.  If it is not, then this will remain an error (though the error message will be better).  You should get this error:

```
foreach statement cannot operate on variables of type 'Range' because 'Range' does not contain a public 
instance or extension definition for 'GetEnumerator'
```
